File: LC/[0053] Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Kadan's algorithm
        # dp[ii]: max subarray sum ending @ ii
        # dp[ii]=max(dp[ii-1]+dp[ii], dp[ii])

        mxs=float('-inf')
        dp=float('-inf')
        for ii in range(len(nums)):
            if ii==0:
                dp=nums[ii]
            else:
                dp=max(dp+nums[ii], nums[ii])

            mxs=max(mxs, dp)

        return mxs

File: LC/[0053] Maximum Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int N = nums.length;
        int mxs=Integer.MIN_VALUE;
        int dp = 0;

        for(var x: nums){
            dp = Math.max(dp+x, x);
            mxs = Math.max(mxs, dp);
        }

        return mxs;
    }
}

/*
* optimize a value from a single array -> type I DP
* -> max(sum(subarray))
* N=10^5 -> NlogN
* dp[ii]: max subarray sum for array ending at ii from nums

*/